pkg/controller/volume/attachdetach/cache/actual_state_of_world.go

```diff
@@ -324,17 +327,20 @@ func (asw *actualStateOfWorld) SetVolumeMountedByNode(
 	if err != nil {
 		return fmt.Errorf("Failed to SetVolumeMountedByNode with error: %v", err)
 	}
+	// TODO: probably safer to switch getNodeAndVolume to return pointers?
 
 	if mounted {
 		// Increment set count
 		nodeObj.mountedByNodeSetCount = nodeObj.mountedByNodeSetCount + 1
 	} else {
 		// Do not allow value to be reset unless it has been set at least once
+		// TODO: not sure I understand the logic here?
```

Comment:
This function is mainly for master keep a record whether the volume is still mounted on the node. This information is updated by the kubelet on the node to the api server and master can get it. 
When volume is attached to the node, in AddVolumeNode function you can see by default mountedByNode is set to true, mountedByNodeSetCount to 0. mountedByNode to true means that the safety check before trying to detach will fail because it is indicating volume is still mounted. 

		volumeObj.nodesAttachedTo[nodeName] = nodeAttachedTo{
			nodeName:              nodeName,
			mountedByNode:         true, // Assume mounted, until proven otherwise
			mountedByNodeSetCount: 0,
			detachRequestedTime:   time.Time{},
		}

  One special logic was put here is that if the current local state showing that volume is not yet mounted to the node (mountedByNodeSetCount=0) we don't set mountedByNode to false. So we want to wait until it is mounted at least once on the node side, so we can mark as umounted so it is safe to detach. @saad-ali add this for some race condition issue existed before.  I am thinking to remove this logic, but only after we think thoroughly about this.